[Fix #2045] Fix error on class method definition with modifier

unmanbearpig opened the pull request:
We were assuming that method body is always the third element of a
method definition node

```
_method_name, _args, body = *args.first
```

and here is an example how args.first looks in case of regular methods:
```
(def :test
  (args)
  (send nil :body))
```

but in case of class methods, for example in this code
```
private_class_method def self.bar
```

there is an extra `(self)` element:
```
(defs
  (self) :bar
  (args)
  (send nil :body))
```

So the body becomes the fourth element instead.

Always taking the last element takes care
of the issue.

bbatsov commented on lib/rubocop/cop/style/indentation_width.rb:
```diff
@@ -85,7 +85,8 @@ def on_send(node)
                                                        args)
 
-          _method_name, _args, body = *args.first
+          body = args.first.to_a.last
```
Generally, I prefer us to use splatting. You can use `*_, body = *args.first` to achieve the same thing.

bbatsov replied:
Alternatively we can check if the node is `def` or `defs`.

unmanbearpig commented:
You're right! I completely forgot that I can use splat on the left side of '=' as well.

bbatsov commented:
Looks good. I'm not sure whether we should add tests include class methods for all indentation style (as we're clearly short on those), but I'm sure these commits have to be squashed before we can merge the PR. :-)

unmanbearpig commented:
I'm not sure about the test as well, do you think it would be better to delete it?

bbatsov commented:
Well, the real problem is not `private_class_method`. The problem is that there are no tests using class methods, so adding a few of those would be beneficial.

unmanbearpig commented:
Good point. I'll see what I can come up with when I get back to my computer. 


> On 19 Jul 2015, at 19:37, Bozhidar Batsov <notifications@github.com> wrote:
> 
> Well, the real problem is not private_class_method. The problem is that there are no tests using class methods, so adding a few of those would be beneficial.
> 
> —
> Reply to this email directly or view it on GitHub.
>

bbatsov commented:
You'll also have to rebase on top of the current `master`.

unmanbearpig commented:
I've deleted my original test and added tests for cases where there was a test for def but not for defs. Not sure if all of them are necessary though. 
What do you think?

I've also verified that my new tests catch the original bug.

bbatsov commented:
:+1: Looks great. Thanks!

unmanbearpig commented:
Awesome! I'm happy to help :)
